domain/removal/service/unit.go

```diff
@@ -107,15 +115,15 @@ func (s *Service) RemoveUnit(
 	unitJobUUID, err := s.unitScheduleRemoval(ctx, unitUUID, force, wait)
 	if err != nil {
 		return "", errors.Capture(err)
-	} else if machineUUID == "" {
-		// If there is no machine UUID, then the unit was not the last one on
-		// the machine, so we can return early.
+	} else if cascaded.IsEmpty() {
+		// No other intities associated with the unit were 
+		// also ensured to be "dying", so we're done.
 		return unitJobUUID, nil
 	}
 
-	s.logger.Infof(ctx, "unit was the last one on machine %q, scheduling removal", machineUUID)
+	s.logger.Infof(ctx, "unit was the last one on machine %q, scheduling removal", *cascaded.MachineUUID)
 
-	s.removeMachines(ctx, []string{machineUUID}, force, wait)
+	s.removeMachines(ctx, []string{*cascaded.MachineUUID}, force, wait)
```

Comment:
There could be a case that storage instances have values, but machine UUID doesn't, which could lead to a panic here. It seems like we could be slightly defensive here and ensure that machine uuid isn't nil.